presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -0,0 +1,178 @@
+        }
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        if (!SystemSessionProperties.isNewOptimizerEnabled(session)) {
+            return plan;
+        }
+
+        long start = System.nanoTime();
+
+        Memo memo = new Memo(idAllocator, plan);
+
+        Lookup lookup = node -> {
+            if (node instanceof GroupReference) {
+                return memo.getExpression(((GroupReference) node).getGroupId());
+            }
+
+            return node;
+        };
+
+        exploreGroup(memo, lookup, idAllocator, symbolAllocator, memo.getRootGroup());
+
+        if (debug) {
+            System.out.println("Groups: " + memo.getGroupCount());
+            System.out.println("Total time: " + Duration.nanosSince(start));
+            System.out.println("Stats:");
+            for (Rule rule : rules) {
+                Stats stat = this.stats.get(rule);
+                System.out.printf("\thits = %3d, time = %8s :: %s\n",
+                        stat.hits,
+                        Duration.succinctNanos(stat.timeNanos),
+                        rule.getClass().getSimpleName());
+            }
+            System.out.println();
+        }
+
+        return memo.extract();
+    }
+
+    private boolean exploreGroup(Memo memo, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, int group)
+    {
+        // tracks whether this group or any children groups change as
+        // this method executes
+        boolean progress = exploreNode(memo, lookup, idAllocator, symbolAllocator, group, memo.getNode(group));
+
+        while (exploreChildren(memo, lookup, idAllocator, symbolAllocator, group)) {
+            progress = true;
+
+            // if children changed, try current group again
+            // in case we can match additional rules
+            if (!exploreNode(memo, lookup, idAllocator, symbolAllocator, group, memo.getNode(group))) {
+                // no additional matches, so bail out
+                break;
+            }
+        }
+
+        return progress;
+    }
+
+    private boolean exploreNode(
+            Memo memo,
+            Lookup lookup,
+            PlanNodeIdAllocator idAllocator,
+            SymbolAllocator symbolAllocator,
+            int group,
+            PlanNode node)
+    {
+        boolean done = false;
+        boolean progress = false;
+
+        while (!done) {
+            done = true;
+            for (Rule rule : rules) {
+                long start = System.nanoTime();
+                Optional<PlanNode> transformed = rule.apply(node, lookup, idAllocator, symbolAllocator);
+                stats.get(rule).timeNanos += System.nanoTime() - start;
+
+                if (transformed.isPresent()) {
+                    stats.get(rule).hits++;
+
+                    memo.transform(group, transformed.get());
```

Comment:
Technically, "addEquivalence" would be more accurate. It's very likely that the Memo implementation will need to be designed to work hand-in-hand with the exploration algorithm, so that would be premature generalization.

`replace` seems reasonable. I'll change it to that.